openassessment/fileupload/backends/gcs.py

```diff
@@ -0,0 +1,73 @@
+"""GCS Bucket File Upload Backend."""
+
+from google.cloud import storage
+import logging
+
+from ..exceptions import FileUploadInternalError
+from .base import BaseBackend
+
+log = logging.getLogger("openassessment.fileupload.api")  # pylint: disable=invalid-name
+
+
+class GCSBackend(BaseBackend):
+    """ GCS Bucket File Upload Backend. """
+
+    def get_upload_url(self, key, content_type):
+        """Get a signed URL for uploading a file to GCS"""
+        bucket_name, key_name = self._retrieve_parameters(key)
+        try:
+            client = _connect_to_gcs()
+            bucket = client.bucket(bucket_name)
+            blob = bucket.blob(key_name)
+            url = blob.generate_signed_url(
+                version="v4",
+                expiration=self.UPLOAD_URL_TIMEOUT,
+                method="PUT",
+                content_type=content_type,
+            )
+            return url
+        except Exception as ex:
+            log.exception("An internal exception occurred while generating an upload URL.")
+            raise FileUploadInternalError(ex) from ex
+
+    def get_download_url(self, key):
+        """Get a signed URL for downloading a file from GCS"""
+        bucket_name, key_name = self._retrieve_parameters(key)
+        try:
+            client = _connect_to_gcs()
+            bucket = client.bucket(bucket_name)
+            blob = bucket.blob(key_name)
+            if not blob.exists():
+                return ""
+            url = blob.generate_signed_url(
+                version="v4",
+                expiration=self.DOWNLOAD_URL_TIMEOUT,
+                method="GET",
+            )
+            return url
+        except Exception as ex:
+            log.exception("An internal exception occurred while generating a download URL.")
+            raise FileUploadInternalError(ex) from ex
```

Comment:
Is there a `FileDownloadInternalError` ?

Reply:
ORA2 code is very old and it's not the best quality. The exception is used all around the code as if it's `ORAFileOperationError` rather than the name you see here

Reply:
I think no, check the super class: https://github.com/correlation-one/edx-ora2/blob/7336ce5e37ec6fa3477d1844c7ebec29309c54eb/openassessment/fileupload/backends/base.py#L62